gcc/rust/expand/rust-macro-expand.cc

```diff
@@ -1072,7 +1075,35 @@ class AttrVisitor : public AST::ASTVisitor
       }
 
     // strip all statements
-    expand_pointer_allow_strip (expr.get_statements ());
+    auto &stmts = expr.get_statements ();
+    for (auto it = stmts.begin (); it != stmts.end ();)
+      {
+	auto &stmt = *it;
+
+	stmt->accept_vis (*this);
+
+	auto fragment = expander.take_expanded_fragment ();
+	if (fragment.is_error ())
+	  {
+	    it++;
+	    continue;
+	  }
+
+	if (!fragment.get_nodes ().empty ())
+	  {
+	    // Remove the current expanded invocation
+	    it = stmts.erase (it);
+	    for (auto &node : fragment.get_nodes ())
+	      {
+		it = stmts.insert (it, AST::SingleASTNode::extract_stmt (node));
```

Comment:
Since `take_expanded_fragment()` takes ownership of the `ASTFragment` using move and the `fragment` isn't used again afterward, wouldn't it be better to `std::move()` each statement from the SingleASTNode instead of cloning it?

This would probably involve rewriting `extract_stmt()` to move and be a member function. 

Also, wouldn't a macro that contained a `SingleASTNode` that wasn't a statement cause an assertion failure here? It should be a fatal error I would think, but assertion failure is probably less useful as an error message to the programmer. Ignore this if there's some check somewhere else that I missed.

Reply:
> Since `take_expanded_fragment()` takes ownership of the `ASTFragment` using move and the `fragment` isn't used again afterward, wouldn't it be better to `std::move()` each statement from the SingleASTNode instead of cloning it?
> 
> This would probably involve rewriting `extract_stmt()` to move and be a member function.

Very good point. This is definitely something that I'll look into, thanks. I think it would be easy to make `extract_stmt` a member function, I had originally made these functions to pass them as lambdas in `expand_pointer_allow_strip`, but that didn't turn out as ergonomic as I'd hoped.

> 
> Also, wouldn't a macro that contained a `SingleASTNode` that wasn't a statement cause an assertion failure here? It should be a fatal error I would think, but assertion failure is probably less useful as an error message to the programmer. Ignore this if there's some check somewhere else that I missed.

This is definitely an issue and the reason why so many tests fail at the moment. The mechanism we have to "prevent" this is that the content of the macro's expansion are parsed according to the current context. So, for example, if we're in a position expecting an expression 
```rust
let a = add!(1, 561, 4);
```
then the parser will only try to parse expressions from the expanded tokens. Otherwise, this is a parse error.

I don't know if this is enough "protection" against the behavior. But ideally, this should never crash for users.

This is also why tests don't pass right now: Since I didn't add visitors to other constructs such as let-assignments, we try to extract statements from something that is clearly an expression if the above example is located in a block.